### PR TITLE
Consumer groups

### PR DESCRIPTION
**Pull request by harlow**

### CODE REVIEW

File: broker.go [hunk 1]
```diff
@@ -31,17 +29,15 @@ func newBroker(
...
+	shardMu    sync.Mutex
+	shards     map[string]*kinesis.Shard
 }
 
-// start is a blocking operation which will loop and attempt to find new
+// Start is a blocking operation which will loop and attempt to find new
 // shards on a regular cadence.
-func (b *broker) start(ctx context.Context) {
-	b.findNewShards()
+func (b *broker) Start(ctx context.Context, shardc chan string) {
```
Comment (harlow) [inline broker.go]: we could potentially adjust this func signature to have it return `chan string`

```go
func (b *broker) Start(ctx context.Context) <-chan string {
```

File: consumergroup.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+// ConsumerGroupCheckpoint is a simple struct for managing the
+type ConsumerGroupCheckpoint struct {
+	Storage           CheckpointStorage
+	kinesis           Kinesis
+	LeaseDuration     time.Duration
+	HeartBeatDuration time.Duration
+	OwnerID           string
+	done              chan struct{}
+	currentLeases     map[string]*Lease //Initially, this will only be one
+	Mutex             *sync.Mutex
```
Comment (harlow) [inline consumergroup.go]: can we use a Mutext Hat here:
https://dmitri.shuralyov.com/idiomatic-go#mutex-hat

Also I think the exported vs non-exported should be grouped together in alphabetical order

File: consumergroup.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+	tick := time.NewTicker(cgc.LeaseDuration)
+	defer tick.Stop()
+	var currentLeases map[string]Lease
+	var previousLeases map[string]Lease
+
+	for {
+		select {
+		case <-tick.C:
+			if len(cgc.currentLeases) == 0 { // only do anything if there are no current leases
```
Comment (harlow) [inline consumergroup.go]: can we reduce indentation and use a guard clause?

```go
if len(cgc.currentLeases) > 0 {
  continue
}

// ...
```

File: consumergroup.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+				fmt.Printf("Attempting to acquire lease for OwnerID=%s\n", cgc.OwnerID)
+				var err error
+				currentLeases, err = cgc.Storage.GetAllLeases()
+				if err != nil {
+					// TODO log this error
+				}
+
+				lease := cgc.CreateOrGetExpiredLease(currentLeases, previousLeases)
+				if lease != nil && lease.LeaseKey != "" {
```
Comment (harlow) [inline consumergroup.go]: can we invert this conditional too? ideally the end of this loop will be the core functionality and not  indented

File: consumergroup/ddb.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/aws/aws-sdk-go/service/dynamodb/expression"
+
+	consumer "github.com/harlow/kinesis-consumer"
```
Comment (harlow) [inline consumergroup/ddb.go]: feels like we might be in danger of cyclical import. I've tried to stick to  general  rule of thump that  sub packages don't know about directory levels above them.

Comment (keperry) [reply consumergroup/ddb.go]: I think we might be fine once we reorganize the packages, but if not, let's revisit this guy.  The only thing it is using right now is the Lease struct.

File: broker.go (hunk 1 above)
Comment (keperry) [reply broker.go]: Let me investigate that.

File: consumergroup.go (hunk 2 above)
Comment (keperry) [reply consumergroup.go]: re-arranged.  This is good idea for all go.  I wish the reformatters in IDEs would do this.

File: consumergroup.go (hunk 3 above)
Comment (keperry) [reply consumergroup.go]: Done

File: consumergroup.go (hunk 4 above)
Comment (keperry) [reply consumergroup.go]: Done

File: consumergroup/ddb.go (hunk 5 above)
Comment (keperry) [reply consumergroup/ddb.go]: I moved Lease to storage - in a new file called lease.go - all storage options should be able to use this.  Let me know if that works.

**Comment by harlow** [comment]
hi @keperry I'm gonna close this PR for now (I can't approve PRs I open). that way you can rebase, and submit a PR from your account .

**Pull request closed without merging**